src/Data/Conduit/Tar/Unix.hs

```diff
@@ -84,6 +84,20 @@ restoreFile FileInfo {..} = do
                     unixModTime = fromInteger (fromIntegral epochInt32)
                 void $ tryAny $ Posix.setSymbolicLinkTimesHiRes fpStr unixModTime unixModTime
 #endif
+        FTHardLink link ->
+            liftIO $ do
+                let linkedFp = decodeFilePath link
+                linkedFileExists <- Posix.fileExist linkedFp
+                -- If the linked file does not exist (yet), we cannot create a hard link.
+                -- Try to "pre-create" it.
+                unless linkedFileExists $ do
+                    Dir.createDirectoryIfMissing True $ Posix.takeDirectory linkedFp
+                    writeFile linkedFp ""
+                Dir.createDirectoryIfMissing True $ Posix.takeDirectory fpStr
+                Posix.createLink linkedFp fpStr
```

Comment:
According to tar format hard links cannot precede the actual files they are pointing to, so "pre-creating" them together with the parent directories is wrong. On the other hand, yet again, the same argument applies as in my other comment above https://github.com/snoyberg/tar-conduit/pull/20/files#r213034851 and we can do those safety precautions whenever untarring is done leniently.
One thing we should do though is to `Posix.removeLink fpStr` if it already exists, cause otherwise we get an error, while default untarring behavior is to overwrite stuff.
@gaborh-da and @ndmitchell What are your thoughts on this?

Reply:
That all makes sense to me - use `removeLink` and use `lenient` to create it. As long as the library is consistent between files and the various types of links then I think any position is reasonable.